### PR TITLE
Fixed test in test_precision_recall_curve and tests/ignite/contrib/metrics/regression

### PR DESCRIPTION
**Pull request by sayantan1410**
Related to #2490 

I couldn't figure out why the test was passing initially, I probably needed to gather data from all the processes with idist.all_gather. But as per the conversation in the mentioned PR I have made the changes to it. Let me know if anything needs to be I will be happy to fix it.

### CODE REVIEW

**Comment by vfdev-5** [comment]
> I couldn't figure out why the test was passing initially, I probably needed to gather data from all the processes with idist.all_gather. 

@sayantan1410 PR seems good. However we need to understand why the tests were passing previously. Please fetch all the data and see why it was passing. Thanks

**Comment by sdesrozis** [comment]
I think there are some others tests broken like this one in regression.

**Comment by sayantan1410** [comment]
@vfdev-5 Yeah trying to look for the reason !!

**Comment by sayantan1410** [comment]
@sdesrozis Should I change them in this PR only ?

**Comment by sdesrozis** [comment]
@sayantan1410 It would be great if you could look for others corrupted tests and fix them if you find any. I would say it's mainly in contrib/metrics/regression. You will just have to rename the PR if others fixes are done.

Thanks a lot for your help !

**Comment by sayantan1410** [comment]
Yeah sure will do that.

**Comment by sayantan1410** [comment]
Probable reason why https://github.com/pytorch/ignite/blob/2eb3bcf4949821a3c894a21a781c5a34bb5545e2/tests/ignite/contrib/metrics/test_precision_recall_curve.py#L196 is passing inspite of being wrong : 
There is a difference is value between precision, recall, thresholds and sk_precision, sk_recall and sk_thresolds in a distributed configuration, however, the difference was small and pytest.approx was taking care of it so that tests were passing. 
Code to reproduce: 
`import torch
import ignite.distributed as idist
from ignite.exceptions import NotComputableError
from ignite.metrics import EpochMetric
from ignite.contrib.metrics import PrecisionRecallCurve
from sklearn.metrics import precision_recall_curve
from ignite.engine import Engine


rank = idist.get_rank()
torch.manual_seed(12)
device = idist.device()
def _test(n_epochs, metric_device):
    metric_device = torch.device(metric_device)
    n_iters = 80
    size = 151
    offset = n_iters * size
    y_true = torch.randint(0, 2, size=(offset * idist.get_world_size(),)).to(device)
    y_preds = torch.randint(0, 2, size=(offset * idist.get_world_size(),)).to(device)
 def update(engine, i):
        return (
            y_preds[i * size + rank * offset : (i + 1) * size + rank * offset],
            y_true[i * size + rank * offset : (i + 1) * size + rank * offset],
        )
    engine = Engine(update)

    prc = PrecisionRecallCurve(device=metric_device)
    prc.attach(engine, "prc")

    data = list(range(n_iters))
    engine.run(data=data, max_epochs=n_epochs)

    assert "prc" in engine.state.metrics

    precision, recall, thresholds = engine.state.metrics["prc"]

    np_y_true = y_true.cpu().numpy().ravel()
    np_y_preds = y_preds.cpu().numpy().ravel()

    sk_precision, sk_recall, sk_thresholds = precision_recall_curve(np_y_true, np_y_preds)
    print(f"precision: {precision}")
    print(f"recall: {recall}")
    print(f"thresholds: {thresholds}")
    print(f"sk_precision: {sk_precision}")
    print(f"sk_recall: {sk_recall}")
    print(f"sk_thresholds: {sk_thresholds}")
metric_devices = ["cpu"]
if device.type != "xla":
    metric_devices.append(idist.device())
for metric_device in metric_devices:
    for _ in range(1):
        print("new test")
        _test(n_epochs=1, metric_device=metric_device)
`

Command I used to run the code: 
`
torchrum --nproc_per_node 2 filename.py --backend prefered backend 
`

**Comment by sayantan1410** [comment]
@vfdev-5 @sdesrozis Hey probably the tests for the metrics in contrib/metrics/regression are fixed, can you check once.

**Comment by sayantan1410** [comment]
@vfdev-5 Yeah, I realized this now. 
Also, can you once review the changes in this PR, it would be helpful for me.

File: ignite/engine/events.py [hunk 1]
```diff
@@ -138,7 +138,7 @@ def __or__(self, other: Any) -> "EventsList":
         return EventsList() | self | other
 
 
-class EventEnum(CallableEventWithFilter, Enum):  # type: ignore[misc]
+class EventEnum(CallableEventWithFilter, Enum):
```
Comment (vfdev-5) [inline ignite/engine/events.py:141]: Can you make a separate PR for that

File: tests/ignite/contrib/metrics/test_precision_recall_curve.py [hunk 2]
```diff
@@ -202,13 +202,15 @@ def _test(n_epochs, metric_device):
...
+        offset = n_iters * size
+        y_true = torch.randint(0, 2, size=(offset * idist.get_world_size(),)).to(device)
+        y_preds = torch.randint(0, 2, size=(offset * idist.get_world_size(),)).to(device)
 
         def update(engine, i):
             return (
-                y_preds[i * size : (i + 1) * size],
-                y_true[i * size : (i + 1) * size],
+                y_preds[i * size + rank * offset : (i + 1) * size + rank * offset],
+                y_true[i * size + rank * offset : (i + 1) * size + rank * offset],
```
Comment (vfdev-5) [inline tests/ignite/contrib/metrics/test_precision_recall_curve.py:213]: Can you explain what that does exactly ?
Have you checked if generated data is different on each process ?

Comment (sdesrozis) [reply tests/ignite/contrib/metrics/test_precision_recall_curve.py:213]: Each process sets the same seed for random so each proc should have the same data (I suppose). However, each proc generates the data for itself and others procs, even if it's similar, each proc works on a different chunk of data. It means no sync over the procs needed to evaluate sklearn. 

Sync is replaced by data duplication (as far I understand).

**Comment by sayantan1410** [comment]
Extremely extremely sorry, I was actually working from my friends machine due to some issue in mine, and by mistake I pushed it from his account's ssh key, taking care of it soon.

**Comment by sayantan1410** [comment]
@vfdev-5 @sdesrozis this was the change I was thinking about, please check once.

File: tests/ignite/contrib/metrics/regression/test_median_absolute_error.py [hunk 3]
```diff
@@ -158,13 +158,15 @@ def _test(n_epochs, metric_device):
         metric_device = torch.device(metric_device)
         n_iters = 80
         size = 105
-        y_true = torch.rand(size=(size,)).to(device)
-        y_preds = torch.rand(size=(size,)).to(device)
+
+        offset = n_iters * size
+        y_true = torch.rand(size=(offset * idist.get_world_size(),)).to(device)
+        y_preds = torch.rand(size=(offset * idist.get_world_size(),)).to(device)
```
Comment (vfdev-5) [inline tests/ignite/contrib/metrics/regression/test_median_absolute_error.py:164]: @sayantan1410 either revert these modifications or make them coherent to others and rename the PR.

File: tests/ignite/contrib/metrics/test_precision_recall_curve.py (hunk 2 above)
Comment (vfdev-5) [reply tests/ignite/contrib/metrics/test_precision_recall_curve.py:211]: @sdesrozis I know what all that does, the question was to @sayantan1410 if he understands what he's doing :)

I agree that it would be better to make tests uniform: `_test_distrib_compute` and `_test_distrib_integration`. We have to open a new issue, explain the problem and send a follow-up PR for that.

File: tests/ignite/contrib/metrics/regression/test_median_absolute_error.py (hunk 3 above)
Comment (sayantan1410) [reply tests/ignite/contrib/metrics/regression/test_median_absolute_error.py:164]: Yeah, making them coherent with others, also some tests are failing, fixing them as well

**Comment by sayantan1410** [comment]
@vfdev-5 yeah, now I understand it better than at that time :) 
Also, what to do to make the tests uniform ?

**Comment by sayantan1410** [comment]
@vfdev-5 Hey, Any idea why the hvd test is failling, other tests are passing now.

**Comment by sayantan1410** [comment]
@vfdev-5 Hey, that is of the old tests, I have made a commit after that and it is showing me that the tests are passing except for hvd. Probably you need to refresh the page once.

**Comment by vfdev-5** [comment]
Oh, I see, sorry. I rerun the tests, output looks strange

File: tests/ignite/contrib/metrics/regression/test_median_relative_absolute_error.py [hunk 4]
```diff
@@ -186,7 +188,7 @@ def update(engine, i):
         e = np.abs(np_y_true - np_y_preds) / np.abs(np_y_true - np_y_true.mean())
         np_res = np.median(e)
 
-        assert pytest.approx(res) == np_res
+        assert pytest.approx(res, rel=1e-3) == np_res
```
Comment (vfdev-5) [inline tests/ignite/contrib/metrics/regression/test_median_relative_absolute_error.py:191]: Why you up the rel tolerance ? If it starts failing than there can be something in the code as well.

File: tests/ignite/contrib/metrics/regression/test_median_absolute_error.py (hunk 3 above)
Comment (vfdev-5) [reply tests/ignite/contrib/metrics/regression/test_median_absolute_error.py:164]: Have you updated others ? Do not forget to rename the PR title to reflect that you change not only ` test_precision_recall_curve.py`

Comment (sayantan1410) [reply tests/ignite/contrib/metrics/regression/test_median_absolute_error.py:164]: Yes, in the other tests this has been changed and I have changed the PR title as well.

Comment (vfdev-5) [reply tests/ignite/contrib/metrics/regression/test_median_absolute_error.py:164]: Why don't you add
```
        rank = idist.get_rank()
        torch.manual_seed(rank)
```
here and other regression files ?

File: tests/ignite/contrib/metrics/regression/test_median_absolute_error.py [hunk 5]
```diff
@@ -185,7 +187,7 @@ def update(engine, i):
         e = np.abs(np_y_true - np_y_preds)
         np_res = np.median(e)
 
-        assert pytest.approx(res) == np_res
+        assert pytest.approx(res, rel=1) == np_res
```
[the remark this answers is not in the document]
Comment (sayantan1410) [reply tests/ignite/contrib/metrics/regression/test_median_absolute_error.py:190]: Yeah I also thought so, I am trying to find why there is this difference in values.

File: tests/ignite/contrib/metrics/regression/test_median_absolute_error.py (hunk 3 above)
Comment (sayantan1410) [reply tests/ignite/contrib/metrics/regression/test_median_absolute_error.py:164]: I added this to precision_recall_curve and was waiting for all the tests to pass for it. Now I will add it to the other files as well.

**Comment by sayantan1410** [comment]
@vfdev-5 Hey, Getting assertion error if I am not increasing the tolerance, Any idea how should I fix this ?

**Comment by vfdev-5** [comment]
@sayantan1410 can we split this PR such that improved tests that are passing could be merged and those are not passing we could investigate case by case. Otherwise you have a stuck block of modifications...

**Comment by sayantan1410** [comment]
@vfdev-5 I could do that, basically, only the precision_recall_curve is working, if possible can we have a meet once, I can show you where I am stuck, and we can solve the issue may be.

**Comment by vfdev-5** [comment]
Closing this PR in favor of #2655. Thanks for your work @sayantan1410 !
